url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/SetupFileParser.java

```diff
@@ -0,0 +1,286 @@
+		
+		String[] requiredPropertiesGeneral = {}; //{ "seed", "useDerby", "printKPIs"};
+		propRequired.put("general_properties", requiredPropertiesGeneral);
+		
+		String[] requiredPropertiesDemographics= { "name", "number_of_entities", "participation_probs_installations", "participation_probs_persons", "participation_probs_apps"};
+		propRequired.put("demographics", requiredPropertiesDemographics); 
+		
+		String[] requiredPropertiesScenario = { "name", "setup"};
+		propRequired.put("scenario", requiredPropertiesScenario);
+		
+		String[] requiredPropertiesSimulation = { "name", "response_type"};
+		propRequired.put("simulation", requiredPropertiesSimulation);
+		
+		String[] requiredPropertiesPricing = { "name", "type", "billingCycle", "fixedCharge"};
+		propRequired.put("pricing_policy", requiredPropertiesPricing);
+		propRequired.put("pricing_policy_baseline", requiredPropertiesPricing);
+		
+		String[] requiredPropertiesPricingTOU = { "timezones", "prices"};
+		propRequired.put("TOUPricing", requiredPropertiesPricingTOU);
+		
+		String[] requiredPropertiesPricingSEP = { "levels", "prices"};
+		propRequired.put("ScalarEnergyPricing", requiredPropertiesPricingSEP);
+		
+		String[] requiredPropertiesPricingSEPTZ = { "offpeakPrice", "offpeakHours", "levels", "prices"};
+		propRequired.put("ScalarEnergyPricingTimeZones", requiredPropertiesPricingSEPTZ);
+		
+		String[] requiredPropertiesPricingEPP = { "contractedCapacity", "energyPrice", "powerPrice"};
+		propRequired.put("EnergyPowerPricing", requiredPropertiesPricingEPP);
+
+		String[] requiredPropertiesPricingMPP = { "energyPrice", "powerPrice"};
+		propRequired.put("MaximumPowerPricing", requiredPropertiesPricingMPP);
+		
+		String[] requiredPropertiesPricingAIP = { "fixedCost", "additionalCost", "contractedEnergy"};
+		propRequired.put("AllInclusivePricing", requiredPropertiesPricingAIP);
+		
+		String[] requiredPropertiesInstallation = { "id", "name"};
+		propRequired.put("installation", requiredPropertiesInstallation);
+		
+		String[] requiredPropertiesPerson = {"id", "name", "installation"};
+		propRequired.put("person", requiredPropertiesPerson);
+		
+		String[] requiredPropertiesConsModel = {"id", "name", "pmodel", "qmodel"};
+		propRequired.put("consumption_model", requiredPropertiesConsModel);
+		
+		String[] requiredPropertiesAppliance = {"id", "name", "installation", "consumption_model"};
+		propRequired.put("appliance", requiredPropertiesAppliance);
+		
+		String[] requiredPropertiesActivity = {"name", "person", "id"};
+		propRequired.put("activity", requiredPropertiesActivity);
+		
+		String[] requiredPropertiesActModel = {"id", "name", "activity", "containsAppliances", "day_type", "duration_distrType", "start_distrType", "repetitions_distrType"};
+		propRequired.put("activity_model", requiredPropertiesActModel);
+		
+	}
+	
+	public void parseFileForProperties(String filename)
+	{
+		try 
+		{
+			BufferedReader br = new BufferedReader(new FileReader(filename));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *br*. The resource is referenced at lines 114, 118. The resource is closed at line 137. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *br* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).